File: server/src/routes/placements.js
const express = require('express');
const { query, queryOne, execute } = require('../db');
const { authRequired, requireRole } = require('../middleware/auth');

const router = express.Router();

// Placement agency: my placements
router.get('/mine', authRequired, requireRole('placement_agency', 'admin'), async (req, res) => {
  try {
    const isAdmin = req.user.role === 'admin';
    let sql = `SELECT p.*, u.name as candidate_name, u.email as candidate_email,
      e.name as employer_name, e.org_name
      FROM placements p
      JOIN users u ON u.id=p.candidate_id
      LEFT JOIN users e ON e.id=p.employer_id`;
    const params = [];
    if (!isAdmin) { sql += ' WHERE p.agency_id=$1'; params.push(req.user.id); }
    sql += ' ORDER BY p.placement_date DESC';
    res.json(await query(sql, params));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Summary stats
router.get('/summary', authRequired, requireRole('placement_agency', 'admin'), async (req, res) => {
  try {
    const agencyId = req.user.id;
    const isAdmin = req.user.role === 'admin';
    const where = isAdmin ? '' : 'WHERE agency_id=$1';
    const params = isAdmin ? [] : [agencyId];
    const yearWhere = isAdmin ? "WHERE EXTRACT(YEAR FROM placement_date)=EXTRACT(YEAR FROM CURRENT_DATE)" : "WHERE agency_id=$1 AND EXTRACT(YEAR FROM placement_date)=EXTRACT(YEAR FROM CURRENT_DATE)";
    const joinedWhere = isAdmin ? "WHERE status='joined'" : "WHERE agency_id=$1 AND status='joined'";

    const total    = await queryOne(`SELECT COUNT(*) c, AVG(ctc) avg_ctc FROM placements ${where}`, params);
    const thisYear = await queryOne(`SELECT COUNT(*) c FROM placements ${yearWhere}`, params);
    const joined   = await queryOne(`SELECT COUNT(*) c FROM placements ${joinedWhere}`, params);

    res.json({
      total: parseInt(total.c || 0),
      avg_ctc: total.avg_ctc ? Math.round(total.avg_ctc) : 0,
      this_year: parseInt(thisYear.c || 0),
      joined: parseInt(joined.c || 0)
    });
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Create placement
router.post('/', authRequired, requireRole('placement_agency', 'admin'), async (req, res) => {
  try {
    const { candidate_id, employer_id, job_title, company, location, ctc, placement_date, status } = req.body;
    if (!candidate_id || !job_title) return res.status(400).json({ error: 'candidate_id and job_title are required' });
    const result = await execute(`
      INSERT INTO placements (agency_id, candidate_id, employer_id, job_title, company, location, ctc, placement_date, status)
      VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9) RETURNING id
    `, [req.user.id, candidate_id, employer_id || null, job_title, company || '', location || '', ctc || null, placement_date || new Date().toISOString().slice(0,10), status || 'placed']);
    res.status(201).json(await queryOne('SELECT * FROM placements WHERE id=$1', [result.rows[0].id]));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Update placement status
router.put('/:id', authRequired, requireRole('placement_agency', 'admin'), async (req, res) => {
  try {
    const pl = await queryOne('SELECT * FROM placements WHERE id=$1', [req.params.id]);
    if (!pl) return res.status(404).json({ error: 'Placement not found' });
    if (req.user.role !== 'admin' && pl.agency_id !== req.user.id)
      return res.status(403).json({ error: 'Not your placement' });
    const { job_title, company, location, ctc, placement_date, status } = req.body;
    await execute(`UPDATE placements SET job_title=$1, company=$2, location=$3, ctc=$4, placement_date=$5, status=$6 WHERE id=$7`, [
      job_title ?? pl.job_title, company ?? pl.company, location ?? pl.location,
      ctc ?? pl.ctc, placement_date ?? pl.placement_date, status ?? pl.status, req.params.id
    ]);
    res.json(await queryOne('SELECT * FROM placements WHERE id=$1', [req.params.id]));
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

// Delete placement
router.delete('/:id', authRequired, requireRole('placement_agency', 'admin'), async (req, res) => {
  try {
    const pl = await queryOne('SELECT * FROM placements WHERE id=$1', [req.params.id]);
    if (!pl) return res.status(404).json({ error: 'Placement not found' });
    if (req.user.role !== 'admin' && pl.agency_id !== req.user.id)
      return res.status(403).json({ error: 'Not your placement' });
    await execute('DELETE FROM placements WHERE id=$1', [req.params.id]);
    res.status(204).end();
  } catch (err) { console.error(err); res.status(500).json({ error: 'Internal server error' }); }
});

module.exports = router;

File: server-python/routers/placements.py
from fastapi import APIRouter, Depends, HTTPException
from database import query, query_one, execute, execute_returning
from auth import auth_required, require_role
from datetime import date

router = APIRouter(prefix="/api/placements", tags=["placements"])


@router.get("/my")
async def candidate_placements(user: dict = Depends(require_role("candidate"))):
    return await query("""SELECT p.*, a.name as agency_name, e.org_name as employer_org
        FROM placements p
        LEFT JOIN users a ON a.id=p.agency_id
        LEFT JOIN users e ON e.id=p.employer_id
        WHERE p.candidate_id=$1 ORDER BY p.placement_date DESC""", [user["id"]])


@router.get("/mine")
async def my_placements(user: dict = Depends(require_role("placement_agency", "admin", "superadmin"))):
    is_admin = user["role"] in ["admin", "superadmin"]
    if is_admin:
        return await query("""SELECT p.*, u.name as candidate_name, u.email as candidate_email,
            e.name as employer_name, e.org_name
            FROM placements p JOIN users u ON u.id=p.candidate_id
            LEFT JOIN users e ON e.id=p.employer_id
            ORDER BY p.placement_date DESC""")
    return await query("""SELECT p.*, u.name as candidate_name, u.email as candidate_email,
        e.name as employer_name, e.org_name
        FROM placements p JOIN users u ON u.id=p.candidate_id
        LEFT JOIN users e ON e.id=p.employer_id
        WHERE p.agency_id=$1 ORDER BY p.placement_date DESC""", [user["id"]])


@router.get("/summary")
async def summary(user: dict = Depends(require_role("placement_agency", "admin", "superadmin"))):
    is_admin = user["role"] in ["admin", "superadmin"]
    cur_year = str(date.today().year)
    if is_admin:
        total = await query_one("SELECT COUNT(*) c, AVG(ctc) avg_ctc FROM placements")
        this_year = await query_one("SELECT COUNT(*) c FROM placements WHERE LEFT(placement_date::text,4)=$1", [cur_year])
        joined = await query_one("SELECT COUNT(*) c FROM placements WHERE status='joined'")
    else:
        aid = user["id"]
        total = await query_one("SELECT COUNT(*) c, AVG(ctc) avg_ctc FROM placements WHERE agency_id=$1", [aid])
        this_year = await query_one("SELECT COUNT(*) c FROM placements WHERE agency_id=$1 AND LEFT(placement_date::text,4)=$2", [aid, cur_year])
        joined = await query_one("SELECT COUNT(*) c FROM placements WHERE agency_id=$1 AND status='joined'", [aid])
    return {
        "total": int(total["c"] or 0),
        "avg_ctc": round(float(total["avg_ctc"])) if total.get("avg_ctc") else 0,
        "this_year": int(this_year["c"] or 0),
        "joined": int(joined["c"] or 0),
    }


@router.post("", status_code=201)
async def create_placement(body: dict, user: dict = Depends(require_role("placement_agency", "admin"))):
    if not body.get("candidate_id") or not body.get("job_title"):
        raise HTTPException(400, detail="candidate_id and job_title are required")
    row = await execute_returning("""INSERT INTO placements
        (agency_id,candidate_id,employer_id,job_title,company,location,ctc,placement_date,status)
        VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9) RETURNING *""",
        [user["id"], body["candidate_id"], body.get("employer_id"),
         body["job_title"], body.get("company", ""), body.get("location", ""),
         body.get("ctc"), body.get("placement_date") or date.today().isoformat(), body.get("status", "placed")])
    return row


@router.put("/{pid}")
async def update_placement(pid: int, body: dict, user: dict = Depends(require_role("placement_agency", "admin"))):
    pl = await query_one("SELECT * FROM placements WHERE id=$1", [pid])
    if not pl: raise HTTPException(404, detail="Placement not found")
    if user["role"] != "admin" and pl["agency_id"] != user["id"]:
        raise HTTPException(403, detail="Not your placement")
    await execute("""UPDATE placements SET job_title=$1,company=$2,location=$3,ctc=$4,placement_date=$5,status=$6 WHERE id=$7""",
        [body.get("job_title") or pl["job_title"], body.get("company") or pl.get("company"),
         body.get("location") or pl.get("location"),
         body.get("ctc") if "ctc" in body else pl.get("ctc"),
         body.get("placement_date") or pl.get("placement_date"),
         body.get("status") or pl.get("status"), pid])
    return await query_one("SELECT * FROM placements WHERE id=$1", [pid])


@router.delete("/{pid}", status_code=204)
async def delete_placement(pid: int, user: dict = Depends(require_role("placement_agency", "admin"))):
    pl = await query_one("SELECT * FROM placements WHERE id=$1", [pid])
    if not pl: raise HTTPException(404, detail="Placement not found")
    if user["role"] != "admin" and pl["agency_id"] != user["id"]:
        raise HTTPException(403, detail="Not your placement")
    await execute("DELETE FROM placements WHERE id=$1", [pid])
